patches/tModLoader/Terraria.ModLoader/ItemLoader.cs

```diff
@@ -439,11 +439,14 @@ private class HookList
 		/// <summary>
 		/// Calls ModItem.PickAmmo, then all GlobalItem.PickAmmo hooks.
 		/// </summary>
-		public static void PickAmmo(Item item, Player player, ref int type, ref float speed, ref int damage, ref float knockback) {
-			item.modItem?.PickAmmo(player, ref type, ref speed, ref damage, ref knockback);
+		public static void PickAmmo(Item shooter, Item item, Player player, ref int type, ref float speed, ref int damage, ref float knockback) {
+			item.modItem?.PickAmmo(shooter, player, ref type, ref speed, ref damage, ref knockback);
+			item.modItem?.PickAmmo(player, ref type, ref speed, ref damage, ref knockback); // deprecated
 
-			foreach (var g in HookPickAmmo.arr)
-				g.Instance(item).PickAmmo(item, player, ref type, ref speed, ref damage, ref knockback);
+			foreach (var g in HookPickAmmo.arr) {
+				g.Instance(item).PickAmmo(shooter, item, player, ref type, ref speed, ref damage, ref knockback);
```

Comment:
Unfortunately, HookPickAmmo is built to only include those items which actually override the hook from the base class, so you'll probably need 2 separate hook lists

Reply:
Oh, i didn't even think about that when i was making this, thanks.

Reply:
Wait, not sure if i understood what you meant, but i realized that i needed two hooklists, one for the new and one for the deprecated PickAmmo methods, because each hooklist only grabs one overload.